src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -258,10 +258,10 @@ public async Task OnConnectionUpAsync(CancellationToken cancellationToken)
         {
             Log.ConnectionUp(_logger, Circuit.Id, Client.ConnectionId);
 
+            await HandlerLock.WaitAsync(cancellationToken);
```

Comment:
`OnConnection*Async` already runs inside of the sync context:

* https://github.com/aspnet/AspNetCore/blob/8d46b3a64ea784c95dddeb9d421c7cda6de993a2/src/Components/Server/src/Circuits/CircuitRegistry.cs#L201-L211
* https://github.com/aspnet/AspNetCore/blob/8d46b3a64ea784c95dddeb9d421c7cda6de993a2/src/Components/Server/src/Circuits/CircuitRegistry.cs#L102
* https://github.com/aspnet/AspNetCore/blob/8d46b3a64ea784c95dddeb9d421c7cda6de993a2/src/Components/Server/src/Circuits/CircuitHost.cs#L222-L230
* https://github.com/aspnet/AspNetCore/blob/8d46b3a64ea784c95dddeb9d421c7cda6de993a2/src/Components/Server/src/Circuits/CircuitHost.cs#L377-L381

This originally ran outside the sync context and we wanted to avoid a scenario where a ConnectionUp \ ConnectionDown would execute concurrently. I think we can simply remove the lock and make an assertion that this is execution inside the sync context.

/cc @rynowak @javiercn

Reply:
Agreed. Using the sync context is the best fix for this. 

I was working on something like this already. @ajaybhargavb - you can cherry pick 70ac08090123e0c0912c6a7f2d7b6c2739e69f64 which has an API you can call.